<h1>
    <center>学习总结</center>
</h1>


# 异常处理

## 一、异常的概念和异常处理的优点

```
面试题：异常和错误的区别？

Error 类和 Exception 类的父类都是 Throwable 类，他们的区别如下。
Error 类一般是指与虚拟机相关的问题，如系统崩溃，虚拟机错误，内存空间不足，方法调用栈溢出等。对于这类
错误的导致的应用程序中断，仅靠程序本身无法恢复和和预防，遇到这样的错误，建议让程序终止。
Exception 类表示程序可以处理的异常，可以捕获且可能恢复。遇到这类异常，应该尽可能处理异常，使程序恢复
运行，而不应该随意终止异常。
Exception 类又分为运行时异常（Runtime Exception）和受检查的异常(Checked Exception )，运行时异
常;ArithmaticException,IllegalArgumentException，编译能通过，但是一运行就终止了，程序不会处理运行时异常，
出现这类异常，程序会终止。而受检查的异常，要么用 try。。。catch 捕获，要么用 throws 字句声明抛出，交给它
的父类处理，否则编译不会通过。
```

异常：程序执行过程中出现的问题，该问题可以由程序本身来克服和恢复。出现异常，则建议修改异常，使程序正常向下执行。

错误：程序执行过程中出现的问题，该问题不能由程序本身来克服和恢复。出现错误，则建议程序停止执行。

优点：

- 提高程序的容错性和健壮性，
- 将业务处理和异常处理代码分离。

## 二、异常体系

![image-20220124142705555](https://gitee.com/DongJinWei/picgo-imgs/raw/master/images/image-20220124142705555.png)

运行时异常：RuntimeException 和直接或间接继承了该类的类。

检查异常：除了运行时异常的其他类。

常见的异常：

1. ArrayIndexOutOfBoundsException（数组是下标越界）
2. NullPoninterException（空指针异常）
3. ArithmeticException（算术异常）
4. MissingResourceException（丢失资源）
5. ClassNotFoundException（找不到类）
6. SQLException（SQL异常）
7. IOException（输入输出异常）
8. ClassCastException（类型转换异常）
9. FileSystemNotFoundException（文件系统没找到异常）
10. NumberFormatException（number格式异常）

## 三、异常处理机制

捕获异常：关键字有 `try, catch, finally`

```java
public class test{
    public static void main(String[] args){
        int a = 1;
        int b = 0;
        
        try {//try监控区域，业务处理
            System.out.println(a/b);
        }catch (ArithmeticException e){//catch（想要捕获的异常类型）捕获异常，异常处理
            System.out.println("程序异常，变量b不能为0");            
        }finally {//处理善后工作，关闭IO流，资源
            System.out.println("finally");
        }
    }
}
```

执行过程：

1. 业务处理没有出现异常，业务处理执行完成后执行 finally 后的语句。
2. 业务处理出现问题，系统抛出一个异常类对象：
   1. 异常类对象被每一个catch 捕获，则执行该 catch 后面的异常处理程序，然后执行 finally 后的语句。程序接着正常执行。
   2. 异常类对象没有被一个 catch 捕获，执行 finally 语句，将异常类对象抛给调用该方法的方法。

注意：

- finally 中的语句始终会被执行。
- tyr 语句中还可以再嵌套try catch finally 语句。
- finally 和 try 中的 return。
- 再最后一个 catch 中捕获 Exception

## 四、抛出异常

throw 异常类对象；

如果该语句没有在try块中，则该语句后的代码不会执行。异常类对象抛给调用该方法的方法。

如果该语句在try块中，则try块中该语句后的代码不会执行。判断是否能够被try块后面的catch捕获。

方法中如果抛出的异常类对象是检查异常类对象，则需要在方法声明位置(即，大括号和小括号中间的位置)使用throws进行声明或使用catch捕获。

throws 异常类名1，异常类名2，异常类名3，异常类名4……

可能会抛出声明的异常类对象。



在调用声明异常的方法时，需要对声明的检查异常进行处理。

处理方法：使用throws进行声明；使用catch捕获；

throw和throws的区别？

代码位置，方法中，大括号和小括号中间。

一个对象和一个或多个类名

一定抛出异常类对象；可能抛出异常类对象；

throw的应用：

## 五、自定义异常类：

class 类名 extends Exception或子类{       //RuntimeException

​    数据成员；

​    方法成员；

}

 

## 六、异常类中常用方法：

getMessage

printStackTrace
